[RFC] Workaround for compilation errors caused by comparison operators returning float instead of bool

yoheiueda opened the pull request:
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

As reported in #227, [torch.ne](https://docs.pytorch.org/docs/stable/generated/torch.ne.html) and other comparison operators return float instead of bool, which can lead to compilation errors.

This PR proposes an interim workaround for this issue. Once `torch.Tensor.to(dtype=...)` is fully implemented on Spyre, this fix can be removed.

**Key changes**:

- torch.Tensor.to(dtype=...) is currently not implemented on Spyre Inductor. This PR makes to(dtype=...) a no-op for Spyre tensors.
- Comparison operators (torch.ne, etc.) are replaced with custom Spyre ops that return float.
- During compilation, the fake op for the original torch.ne is skipped, and the fake op for the custom Spyre op is used instead.

This approach avoids dtype mismatch errors during compilation.

**Concerns**:
torch.Tensor.to is used for various purposes, so overriding it requires caution. If there is a more targeted hook for data conversion, that would be preferable.


#### Which issue(s) this PR is related to:

Workaround for:
- #227 

Completes the following PRs:
- #65 
- #155
- #106 
- #269 
- #270 
- #271 

Related to:
- #39

#### Special notes for your reviewer:


#### Does this PR introduce a user-facing change?

- Spyre tensors ignore to(dtype=...) requests (no-op)
- Comparison ops on Spyre return float instead of bool

#### Additional note:

This PR allows us to compile and run the following code.
```
import torch
import torch.nn as nn

class test(nn.Module):
    def forward(self, input1, input2, input3):
        tmp = torch.ne(input1, input2)
        tmp = tmp.to(dtype=torch.float16)
        return torch.matmul(tmp, input3)

input1_cpu = torch.ones(64, 64, dtype=torch.float16)
input2_cpu = torch.zeros(64, 64, dtype=torch.float16)
input3_cpu = torch.ones(64, 64, dtype=torch.float16)

input1_spyre = input1_cpu.to("spyre")
input2_spyre = input2_cpu.to("spyre")
input3_spyre = input3_cpu.to("spyre")

net = test().to("spyre")
net_compile = torch.compile(net, backend="inductor")
output_spyre = net_compile(input1_spyre, input2_spyre, input3_spyre)
output_cpu = output_spyre.cpu()

print(f"Spyre result\n{output_cpu}")
```

This code failed with:

```
  File "/home/yohei/dt-inductor/pytorch/torch/_inductor/ops_handler.py", line 132, in to_dtype
    raise NotImplementedError
torch._inductor.exc.InductorError: NotImplementedError:
```

And when the line of `to(dtype=...)` is commented out:

```
  File "/home/yohei/dt-inductor/pytorch/torch/_inductor/pattern_matcher.py", line 1339, in check_fn
    args[i] = torch.empty_strided(
              ^^^^^^^^^^^^^^^^^^^^
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RuntimeError: Only Tensors of floating point and complex dtype can require gradients
```

dgrove-oss commented:
Correct. for going from boolean to float16 no new DDL is needed.   We just need to augment the SDSC generation to generate the right datadsc.  Avery has the code on a branch / git history.  We just need to recover it.

thoangtrvn commented:
I'm working on the code change for #83  now.

yoheiueda commented:
@dgrove-oss @thoangtrvn @avery-blanchard Thank you very much for the PRs! I’ve verified that the comparison operators now correctly return bool tensors.

I’ve updated this PR with a few additional changes. Could you please take a look?

* Added support for torch.gt and torch.ne via decomposition
* Added test cases
* Removed float32 input support

```
(dti-venv) [yohei@yohei-spyre-dev torch-spyre]$ pytest ./tests
================================================ test session starts ================================================
platform linux -- Python 3.12.9, pytest-9.0.1, pluggy-1.6.0
rootdir: /home/yohei/dt-inductor/torch-spyre
configfile: pyproject.toml
plugins: hypothesis-6.148.2
collected 233 items

tests/_inductor/test_building_blocks.py .s.s..                                                                [  2%]
tests/_inductor/test_inductor_ops.py ..s..................................................................... [ 33%]
................................................................                                              [ 60%]
tests/tensor/test_tensor_layout.py ........                                                                   [ 64%]
tests/test_fallbacks.py ........                                                                              [ 67%]
tests/test_modules.py ssssss.                                                                                 [ 70%]
tests/test_ops.py ..ssssss...................sss..ssss.s.......ss.ss                                          [ 92%]
tests/test_spyre.py .s...s...........                                                                         [ 99%]
tests/test_spyre_lazy_silent.py .                                                                             [100%]

================================================= warnings summary ==================================================
tests/_inductor/test_inductor_ops.py::TestOps::test_activation_cls_gelu_fp16
  /home/yohei/dt-inductor/pytorch/torch/_inductor/ops_handler.py:745: UserWarning: undefined OpHandler.gelu, please add missing op schema
    warnings.warn(f"undefined OpHandler.{name}, please add missing op schema")

tests/_inductor/test_inductor_ops.py::TestOps::test_pointwise_range_op_clamp_fp16
  /home/yohei/dt-inductor/pytorch/torch/_inductor/ops_handler.py:745: UserWarning: undefined OpHandler.clamp, please add missing op schema
    warnings.warn(f"undefined OpHandler.{name}, please add missing op schema")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================== 204 passed, 29 skipped, 2 warnings in 85.10s (0:01:25) ===============================
```
